base/compiler/typeinfer.jl

```diff
@@ -340,14 +340,16 @@ function CodeInstance(
         relocatability)
 end
 
-function maybe_compress_codeinfo(interp::AbstractInterpreter, linfo::MethodInstance, ci::CodeInfo)
-    def = linfo.def
+function maybe_compress_codeinfo(interp::AbstractInterpreter, mi::MethodInstance, ci::CodeInfo)
+    def = mi.def
     toplevel = !isa(def, Method)
     if toplevel
         return ci
     end
     if may_discard_trees(interp)
-        cache_the_tree = ci.inferred && (is_inlineable(ci) || isa_compileable_sig(linfo.specTypes, linfo.sparam_vals, def))
+        cache_the_tree = ci.inferred && (
+            is_inlineable(interp, ci, InlineeMetaInfo(ci.rettype, mi)) ||
+            isa_compileable_sig(mi.specTypes, mi.sparam_vals, def))
```

Comment:
Why do we want to cache this non-inlineable source that is `compileable_sig`? Maybe @vtjnash or @Keno ?